### PR TITLE
Fix #15463: use intersection types to emulate spread in generic react  components

### LINKED ISSUE #15463: is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes
**TypeScript Version:**  2.3.1

**Code**

With the above mentioned version I can't spread Props objects anymore to child React components, when the Props don't match up 100%.

```ts

/* App.tsx */

interface Props {
    ...
    scopes: any;
}

interface State {}

export class App extends React.Component<Props, State> {
    render() {
         return <Content {...this.props} />; // error
    }
}

/* Content.tsx */

interface Props {
    ...
    // without scopes in Props
}

export default class Content extends React.Component<Props, {}>{
    ...
}
```

**Expected behavior:**
No error and there wasn't any on 2.2.1

**Actual behavior:**

error:
```
ERROR in ./spa/js/comps/App.tsx
(121,49): error TS2322: Type '{ ui?: UIState; session?: SessionState; scopes?: any; users?: any; roles?: any; plugins?: any; us...' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Content> & Props & { children?: ReactNode; }'.
  Property 'scopes' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Content> & Props & { children?: ReactNode; }'.
```

### PR DESCRIPTION
**Pull request by mhegazy**
Fixes #15463

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -13404,7 +13410,13 @@ namespace ts {
...
             }
 
-            return hasSpreadAnyType ? anyType : createJsxAttributesType(attributes.symbol, attributesTable);
+            if (hasSpreadAnyType) {
+                return anyType;
+            }
+
+            const attributeType = createJsxAttributesType(attributes.symbol, attributesTable);
+            return typeToIntersect && attributesTable.size ? getIntersectionType([typeToIntersect, attributeType]) :
+                typeToIntersect ? typeToIntersect : attributeType;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: How about a readable form?

```ts
if (typeToIntersect) {
    if (attributesTable.size) {
        return getIntersectionType([typeToIntersect, attributeType]);
    }
    return typeToIntersect;
}
return attributeType;
```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -13345,11 +13346,16 @@ namespace ts {
                         attributesArray = [];
                         attributesTable = createMap<Symbol>();
                     }
-                    const exprType = getApparentType(checkExpression(attributeDecl.expression));
+                    const exprType = checkExpression(attributeDecl.expression);
                     if (isTypeAny(exprType)) {
                         hasSpreadAnyType = true;
                     }
-                    spread = getSpreadType(spread, exprType);
+                    if (!isValidSpreadType(exprType)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Invert the condition here and flip them around 😄

File: src/compiler/checker.ts (hunk 1 above)
Comment (mhegazy) [inline src/compiler/checker.ts]: We use the ternary operator quite often. so do not think this is out of line here.

Comment (alfaproject) [inline src/compiler/checker.ts]: You would save one condition and make it way easier to read in this case tho

**Pull request merged**